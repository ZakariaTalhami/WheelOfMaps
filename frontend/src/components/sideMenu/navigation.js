import { FaFileAlt, FaCog, FaHandPointer } from "react-icons/fa";
import PropTypes from "prop-types";
import EntityViewer from "./entityViews/EntityViewer";
import ChapterSummary from "./ChapterSummary";
import NotImplementedMessage from "./NotImplementedMessage";

export const CHAPTER_SUMMARY = "chapter-summary";
export const SELECTED_SUMMARY = "selected-summary";
export const PIZZA = "PIZZA";
export const SETTINGS = "settings";

const getDevNavigation = () => {
    let devNavigation = [];

    if (process.env.NODE_ENV === "development") {
        devNavigation = [];
    }

    return devNavigation;
};

export const NAVIGATION = [
    {
        name: CHAPTER_SUMMARY,
        label: "Chapter Summary",
        Icon: FaFileAlt,
        Content: ChapterSummary,
    },
    {
        name: SELECTED_SUMMARY,
        label: "Selected Summary",
        Icon: FaHandPointer,
        Content: EntityViewer,
    },
    {
        name: SETTINGS,
        label: "Settings",
        Icon: FaCog,
        Content: NotImplementedMessage,
    },
    ...getDevNavigation(),
];

export const NavigationType = PropTypes.shape({
    name: PropTypes.string,
    label: PropTypes.string,
    Icon: PropTypes.elementType,
    Content: PropTypes.elementType,
});
