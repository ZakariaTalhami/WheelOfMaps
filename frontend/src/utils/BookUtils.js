/*
    Uility module for common functionality for Books and chapters.
*/

export const getBookChapterTitles = (book) => {
    const chapters = book.chapters || [];
    return chapters.map((chapter) => {
        const prefix =
            chapter.number === 0 ? "Prologue" : `Chapter ${chapter.number}`;

        return `${prefix}: ${chapter.title}`;
    });
};
