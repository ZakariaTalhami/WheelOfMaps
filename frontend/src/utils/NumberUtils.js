/**
 * Pads a number depending on the number of digits required
 * Returns a string with the number padded with zeros
 *
 * @param {Number} num - the desired number
 * @param {Number} places - number of digits
 * @returns {String} formated string with zero padding
 */
export function zeroPad(num, places) {
    var zero = places - num.toString().length + 1;
    return Array(+(zero > 0 && zero)).join("0") + num;
}
