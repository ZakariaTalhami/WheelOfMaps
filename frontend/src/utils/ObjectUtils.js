import _ from "lodash";

const evaluateObjectPath = (obj, path = [], defaultValue) => {
    if (_.isString(path)) path = _.split(path, ".");
    if (!_.isObject(obj) || _.isEmpty(path)) return obj || defaultValue;

    return evaluateObjectPath(obj[_.head(path)], _.drop(path), defaultValue);
};

const ObjectUtils = {
    evaluateObjectPath,
};

export default ObjectUtils;
