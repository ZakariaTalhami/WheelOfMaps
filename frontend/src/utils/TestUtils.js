import React from "react";
import { render } from "@testing-library/react";
import { ChakraProvider } from "@chakra-ui/react";
//Global Theme
import extendedTheme from "../assets/styles/extendedTheme";
// TODO: remove custom font icons, replace with react-icons
import "../assets/fonts/style.css";
// Leafletjs Styles
import "leaflet/dist/leaflet.css";
import "../assets/_leafletjs.override.scss";

const AllTheProviders = ({ children }) => {
    return <ChakraProvider theme={extendedTheme}>{children}</ChakraProvider>;
};

const customRender = (ui, options) =>
    render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
