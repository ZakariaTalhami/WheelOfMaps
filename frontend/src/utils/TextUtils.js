import { Text } from "@chakra-ui/react";

/**
 * Split the summary by new line and wrap in paragraphs
 * @param {String} summary
 * @returns {Array} paragraphs of the summary
 */
export const getSummaryParagraphs = (summary) => {
    const paragraphs = summary.split("\r\n\r\n");
    return paragraphs.map((para, index) => (
        <Text
            w="100%"
            role="paragraph"
            style={{ textIndent: "1.5rem" }}
            key={index}
        >
            {para}
        </Text>
    ));
};
