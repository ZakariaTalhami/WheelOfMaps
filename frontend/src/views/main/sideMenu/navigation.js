import PropTypes from "prop-types";
// Components
import {
    FaFileAlt,
    FaCog,
    FaHandPointer,
    FaPen,
    FaBookMedical,
    FaBookOpen,
} from "react-icons/fa";
import EntityViewer from "./entityViews/EntityViewer";
import NotImplementedMessage from "./NotImplementedMessage";
import ChapterSummary from "./ChapterSummary";
import ChapterForm from "./developementEditor/EntityForms/ChapterForm";
import BookForm from "./developementEditor/EntityForms/BookForm";

export const CHAPTER_FORM = "chapter-form";
export const BOOK_FORM = "book-form";
export const SELECTED_FORM = "selected-form";
 
const getDevNavigation = () => {
    let devNavigation = [];

    if (process.env.NODE_ENV === "development") {
        devNavigation = [
            {
                name: SELECTED_FORM,
                label: "Selected Form",
                Icon: FaPen,
                Content: () => "Selected Form",
            },
            {
                name: BOOK_FORM,
                label: "Book Form",
                Icon: FaBookMedical,
                Content: ChapterForm,
            },
            {
                name: CHAPTER_FORM,
                label: "Chapter Form",
                Icon: FaBookOpen,
                Content: BookForm,
            },
        ];
    }

    return devNavigation;
};

export const CHAPTER_SUMMARY = "chapter-summary";
export const SELECTED_SUMMARY = "selected-summary";
export const SETTINGS = "settings";

export const NAVIGATION = [
    {
        name: CHAPTER_SUMMARY,
        label: "Chapter Summary",
        Icon: FaFileAlt,
        Content: ChapterSummary,
    },
    {
        name: SELECTED_SUMMARY,
        label: "Selected Summary",
        Icon: FaHandPointer,
        Content: EntityViewer,
    },
    {
        name: SETTINGS,
        label: "Settings",
        Icon: FaCog,
        Content: NotImplementedMessage,
    },
    ...getDevNavigation(),
];

export const NavigationType = PropTypes.shape({
    name: PropTypes.string,
    label: PropTypes.string,
    Icon: PropTypes.elementType,
    Content: PropTypes.elementType,
});
