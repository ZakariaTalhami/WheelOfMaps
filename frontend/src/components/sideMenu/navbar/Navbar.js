import { Box, Flex, Icon, Tooltip, VStack } from "@chakra-ui/react";
import React from "react";
import PropTypes from "prop-types";
import { NavigationType } from "../navigation";

const NavbarWraper = (props) => (
    <Flex
        pos="relative"
        justify="center"
        bgColor="primaryColor"
        height="100%"
        w="50px"
        shadow="4px -4px 10px 4px rgba(0, 0, 0, 0.25);"
        {...props}
    />
);

const NavIcon = (props) => (
    <Tooltip
        hasArrow
        label={props.label}
        placement="right"
        bg="neutralColor"
        color="primaryColor"
    >
        {/* Added a Box to fix the forwardRef issue with react-icons */}
        <Box
            borderRight="4px solid"
            borderColor={props.isSelected ? "neutralColor" : "transparent"}
            textAlign="center"
            w="100%"
            p="10px 0"
        >
            <Icon
                cursor="pointer"
                boxSize="35px"
                color="neutralColor"
                {...props}
            />
        </Box>
    </Tooltip>
);

const Navbar = ({ onSelect, selected, navigation }) => {
    return (
        <NavbarWraper>
            <VStack spacing="20px" w="100%" pt="25px">
                {navigation.map((opt) => (
                    <NavIcon
                        key={opt.name}
                        onClick={() => onSelect(opt.name)}
                        isSelected={selected === opt.name}
                        label={opt.label}
                        as={opt.Icon}
                    />
                ))}
            </VStack>
        </NavbarWraper>
    );
};

Navbar.propTypes = {
    selected: PropTypes.string,
    navigation: PropTypes.arrayOf(NavigationType),
};

export default Navbar;
