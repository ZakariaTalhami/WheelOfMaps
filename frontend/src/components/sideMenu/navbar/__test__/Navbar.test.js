import React from "react";
import { fireEvent, render, screen } from "../../../../utils/TestUtils";
import { NAVIGATION } from "../../navigation";
import Navbar from "../Navbar";

test("renders all the navigation links", () => {
    render(<Navbar navigation={NAVIGATION} />);

    expect(screen.getAllByRole("link").length).toEqual(NAVIGATION.length);
});

test("Hover over link renders tooltip", async () => {
    render(<Navbar navigation={NAVIGATION} />);

    const links = screen.getAllByRole("link");
    fireEvent.mouseOver(links[0]);

    await screen.findByRole("tooltip");
    expect(screen.getByText(NAVIGATION[0].label)).toBeInTheDocument();
});

test("Navigation shows selection indicator", () => {
    const selectedNavigation = NAVIGATION[0].name;
    render(<Navbar navigation={NAVIGATION} selected={selectedNavigation} />);

    expect(screen.getAllByRole("link")[0]).toHaveStyle(
        "border-color: var(--chakra-colors-neutralColor)"
    );
});

test("Clicking triggers selection", () => {
    const onSelect = jest.fn();
    render(<Navbar navigation={NAVIGATION} onSelect={onSelect} />);

    const links = screen.getAllByRole("link");
    fireEvent.click(links[0]);

    expect(onSelect).toHaveBeenCalledWith(NAVIGATION[0].name);

    fireEvent.click(links[1]);

    expect(onSelect).toHaveBeenCalledWith(NAVIGATION[1].name);
});

test("matches snapshot - no selection", () => {
    const { container } = render(<Navbar navigation={NAVIGATION} />);

    expect(container).toMatchSnapshot();
});

test("matches snapshot - with selection", () => {
    const selectedNavigation = NAVIGATION[0].name;
    const { container } = render(
        <Navbar navigation={NAVIGATION} selected={selectedNavigation} />
    );

    expect(container).toMatchSnapshot();
});
