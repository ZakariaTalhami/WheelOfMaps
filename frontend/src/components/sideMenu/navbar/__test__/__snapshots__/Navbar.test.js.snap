// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot - no selection 1`] = `
<div>
  <div
    class="css-1ficosa"
  >
    <div
      class="chakra-stack css-1kixdby"
    >
      <div
        class="css-1ez7mdk"
        role="link"
      >
        <svg
          class="chakra-icon css-16c6au9"
          fill="currentColor"
          focusable="false"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 384 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M224 136V0H24C10.7 0 0 10.7 0 24v464c0 13.3 10.7 24 24 24h336c13.3 0 24-10.7 24-24V160H248c-13.2 0-24-10.8-24-24zm64 236c0 6.6-5.4 12-12 12H108c-6.6 0-12-5.4-12-12v-8c0-6.6 5.4-12 12-12h168c6.6 0 12 5.4 12 12v8zm0-64c0 6.6-5.4 12-12 12H108c-6.6 0-12-5.4-12-12v-8c0-6.6 5.4-12 12-12h168c6.6 0 12 5.4 12 12v8zm0-72v8c0 6.6-5.4 12-12 12H108c-6.6 0-12-5.4-12-12v-8c0-6.6 5.4-12 12-12h168c6.6 0 12 5.4 12 12zm96-114.1v6.1H256V0h6.1c6.4 0 12.5 2.5 17 7l97.9 98c4.5 4.5 7 10.6 7 16.9z"
          />
        </svg>
      </div>
      <div
        class="css-1ez7mdk"
        role="link"
      >
        <svg
          class="chakra-icon css-16c6au9"
          fill="currentColor"
          focusable="false"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M448 240v96c0 3.084-.356 6.159-1.063 9.162l-32 136C410.686 499.23 394.562 512 376 512H168a40.004 40.004 0 0 1-32.35-16.473l-127.997-176c-12.993-17.866-9.043-42.883 8.822-55.876 17.867-12.994 42.884-9.043 55.877 8.823L104 315.992V40c0-22.091 17.908-40 40-40s40 17.909 40 40v200h8v-40c0-22.091 17.908-40 40-40s40 17.909 40 40v40h8v-24c0-22.091 17.908-40 40-40s40 17.909 40 40v24h8c0-22.091 17.908-40 40-40s40 17.909 40 40zm-256 80h-8v96h8v-96zm88 0h-8v96h8v-96zm88 0h-8v96h8v-96z"
          />
        </svg>
      </div>
      <div
        class="css-1ez7mdk"
        role="link"
      >
        <svg
          class="chakra-icon css-16c6au9"
          fill="currentColor"
          focusable="false"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M487.4 315.7l-42.6-24.6c4.3-23.2 4.3-47 0-70.2l42.6-24.6c4.9-2.8 7.1-8.6 5.5-14-11.1-35.6-30-67.8-54.7-94.6-3.8-4.1-10-5.1-14.8-2.3L380.8 110c-17.9-15.4-38.5-27.3-60.8-35.1V25.8c0-5.6-3.9-10.5-9.4-11.7-36.7-8.2-74.3-7.8-109.2 0-5.5 1.2-9.4 6.1-9.4 11.7V75c-22.2 7.9-42.8 19.8-60.8 35.1L88.7 85.5c-4.9-2.8-11-1.9-14.8 2.3-24.7 26.7-43.6 58.9-54.7 94.6-1.7 5.4.6 11.2 5.5 14L67.3 221c-4.3 23.2-4.3 47 0 70.2l-42.6 24.6c-4.9 2.8-7.1 8.6-5.5 14 11.1 35.6 30 67.8 54.7 94.6 3.8 4.1 10 5.1 14.8 2.3l42.6-24.6c17.9 15.4 38.5 27.3 60.8 35.1v49.2c0 5.6 3.9 10.5 9.4 11.7 36.7 8.2 74.3 7.8 109.2 0 5.5-1.2 9.4-6.1 9.4-11.7v-49.2c22.2-7.9 42.8-19.8 60.8-35.1l42.6 24.6c4.9 2.8 11 1.9 14.8-2.3 24.7-26.7 43.6-58.9 54.7-94.6 1.5-5.5-.7-11.3-5.6-14.1zM256 336c-44.1 0-80-35.9-80-80s35.9-80 80-80 80 35.9 80 80-35.9 80-80 80z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`matches snapshot - with selection 1`] = `
<div>
  <div
    class="css-1ficosa"
  >
    <div
      class="chakra-stack css-1kixdby"
    >
      <div
        class="css-shin7"
        role="link"
      >
        <svg
          class="chakra-icon css-16c6au9"
          fill="currentColor"
          focusable="false"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 384 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M224 136V0H24C10.7 0 0 10.7 0 24v464c0 13.3 10.7 24 24 24h336c13.3 0 24-10.7 24-24V160H248c-13.2 0-24-10.8-24-24zm64 236c0 6.6-5.4 12-12 12H108c-6.6 0-12-5.4-12-12v-8c0-6.6 5.4-12 12-12h168c6.6 0 12 5.4 12 12v8zm0-64c0 6.6-5.4 12-12 12H108c-6.6 0-12-5.4-12-12v-8c0-6.6 5.4-12 12-12h168c6.6 0 12 5.4 12 12v8zm0-72v8c0 6.6-5.4 12-12 12H108c-6.6 0-12-5.4-12-12v-8c0-6.6 5.4-12 12-12h168c6.6 0 12 5.4 12 12zm96-114.1v6.1H256V0h6.1c6.4 0 12.5 2.5 17 7l97.9 98c4.5 4.5 7 10.6 7 16.9z"
          />
        </svg>
      </div>
      <div
        class="css-1ez7mdk"
        role="link"
      >
        <svg
          class="chakra-icon css-16c6au9"
          fill="currentColor"
          focusable="false"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M448 240v96c0 3.084-.356 6.159-1.063 9.162l-32 136C410.686 499.23 394.562 512 376 512H168a40.004 40.004 0 0 1-32.35-16.473l-127.997-176c-12.993-17.866-9.043-42.883 8.822-55.876 17.867-12.994 42.884-9.043 55.877 8.823L104 315.992V40c0-22.091 17.908-40 40-40s40 17.909 40 40v200h8v-40c0-22.091 17.908-40 40-40s40 17.909 40 40v40h8v-24c0-22.091 17.908-40 40-40s40 17.909 40 40v24h8c0-22.091 17.908-40 40-40s40 17.909 40 40zm-256 80h-8v96h8v-96zm88 0h-8v96h8v-96zm88 0h-8v96h8v-96z"
          />
        </svg>
      </div>
      <div
        class="css-1ez7mdk"
        role="link"
      >
        <svg
          class="chakra-icon css-16c6au9"
          fill="currentColor"
          focusable="false"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M487.4 315.7l-42.6-24.6c4.3-23.2 4.3-47 0-70.2l42.6-24.6c4.9-2.8 7.1-8.6 5.5-14-11.1-35.6-30-67.8-54.7-94.6-3.8-4.1-10-5.1-14.8-2.3L380.8 110c-17.9-15.4-38.5-27.3-60.8-35.1V25.8c0-5.6-3.9-10.5-9.4-11.7-36.7-8.2-74.3-7.8-109.2 0-5.5 1.2-9.4 6.1-9.4 11.7V75c-22.2 7.9-42.8 19.8-60.8 35.1L88.7 85.5c-4.9-2.8-11-1.9-14.8 2.3-24.7 26.7-43.6 58.9-54.7 94.6-1.7 5.4.6 11.2 5.5 14L67.3 221c-4.3 23.2-4.3 47 0 70.2l-42.6 24.6c-4.9 2.8-7.1 8.6-5.5 14 11.1 35.6 30 67.8 54.7 94.6 3.8 4.1 10 5.1 14.8 2.3l42.6-24.6c17.9 15.4 38.5 27.3 60.8 35.1v49.2c0 5.6 3.9 10.5 9.4 11.7 36.7 8.2 74.3 7.8 109.2 0 5.5-1.2 9.4-6.1 9.4-11.7v-49.2c22.2-7.9 42.8-19.8 60.8-35.1l42.6 24.6c4.9 2.8 11 1.9 14.8-2.3 24.7-26.7 43.6-58.9 54.7-94.6 1.5-5.5-.7-11.3-5.6-14.1zM256 336c-44.1 0-80-35.9-80-80s35.9-80 80-80 80 35.9 80 80-35.9 80-80 80z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
